### ISSUE TITLE
setNotificationOpenedHandler does not executes when app is terminated/not opened at all

### ISSUE THREAD

**Issue by apoorvpandey0** [OneSignal/OneSignal-Flutter-SDK#540]
setNotificationOpenedHandler works very well when app is in background or in focus, but when app is swiped away i.e terminated by the user, This listener is not being called so I'm not able to access notification additional data.
I'm testing this with Flutter 2.10.1 and Onesginal SDK 3.2.8
I hope I'm clear, please help!

**Comment by andikanugraha** [comment]
yes, I'm also having this issue. Is there any solution for this?

**Comment by mjarpitanand** [comment]
Same error i am getting.

**Comment by ayechanmaung47** [comment]
We all get errors like that.

**Comment by mjarpitanand** [comment]
Any solution for this

**Comment by AmmarHazem** [comment]
Same issue here, any update on this ?

**Comment by nan-li** [comment]
Hi all,

Thank you for reporting this issue.

Can you share the device information for this? Is it on Android and/or iOS?

**Comment by AmmarHazem** [comment]
I got this issue on iOS iPhone XR

**Comment by apoorvpandey0** [comment]
> Hi all,
> 
> Thank you for reporting this issue.
> 
> Can you share the device information for this? Is it on Android and/or iOS?

For me it was on Android

**Comment by marcorubagotti** [comment]
Any updates?

**Comment by emawby** [comment]
@marcorubagotti We are still investigating the issue and working on a fix. I apologize for the delay and thank you for your patience

**Comment by mrqaidi** [comment]
any news on this ?

**Comment by daveshirman** [comment]
Of course it's not executed, the app isn't even loaded. To workaround:

1. You need to store the result of the notification when your app loads e.g. in local storage, inside this method: `setNotificationOpenedHandler`.

2. Then at some point after, check your storage for any notification data that came in.

After the app is loaded, the handler will run as expected.

**Comment by alexobviously** [comment]
I also have this issue and the workaround above does not work, as the notification handler (passed to `setNotificationOpenedHandler` is never called if the app is opened from a notification.

Did this ever work? Surely this is core functionality?

**Comment by daveshirman** [comment]
Not sure what to tell you, mine works doing exactly what I said.

Whether it's a cold boot from a notification, where the app starts from
zero, or whether it's already in memory, I get my data through in the
handler and everything is working as expected.

Can paste some code this week if that'd give an indication of how mine
works?

**Comment by daveshirman** [comment]
Just thought, you're not running AdBlock or anything similar on the device?

I've had issues before when that was enabled.

Also, are you setting up the handler in main.dart before MaterialApp gets
executed?

**Comment by alexobviously** [comment]
Yeah I have all the onesignal initialisation stuff before `MaterialApp`, even before `WidgetsFlutterBinding.ensureInitialized();`, and `OneSignal.shared.setAppId` is awaited. I've tried moving things around, adding delays in, etc, but it seems like nothing works. To confirm that the handler is never called, I put an exception in it, and it never occurs. The handler works as expected in all other cases - it's just when the app is opened from a notification that it doesn't.

Nothing like AdBlock running (though that is an interesting idea).

@daveshirman if you do have code you could share it would be very much appreciated (no rush!), maybe I've missed something obvious.

**Comment by daveshirman** [comment]
What about after WidgetsFlutterBinding.ensureInitialized()

?

**Comment by daveshirman** [comment]
Hey, so here's how I do it, I've reduced the code where possible to make it simple.


1) I initialise the FlutterBinding stuff..
2) I await setup local key value storage with `GetStorage`.
3) I register OneSignal, with my custom class `OneSignalUtils`
4) I wrap my app in a container widget" `App` inside `MaterialApp`

Notes:

- If the notification is tapped when the app isn't loaded, my app loads.
- It registers OneSignal.
- Then the handler will check if there's a callback registered that pertains to the additionalData.
- If there's no callback set, the value is saved to storage. Otherwise the callback is called.
- `App()` loads up and sets those callbacks in `OneSignalUtils`.  This way, the next time, they'll be called.
- Pay attention to the comments in method: `_checkForPushNotificationOnLoad()`

You can see how I optionally load one of 2 pages in my app - Community or Calendar.

All this code works and is in production in several apps.  It may not be the best way, but it works for me. Please don't reply telling my not to use singletons or any other code advice, thanks!

Please let me know if this all helps :)

### Code....

### In main.ts:

```
void main() async {
  WidgetsFlutterBinding.ensureInitialized();

  await setupLocalStorage();
  OneSignalUtils.init();

  runApp(
    MyApp(),
  );
}

Future<bool> setupLocalStorage() async {
  return GetStorage.init();
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    LoginResponse? user = StorageService().getUser();
    bool loggedIn = (user != null && user.token.isNotEmpty);

    return MaterialApp(
      title: "My app!",
      home: App()
    );
  }
}
```

### In App - the wrapper, just a stateful widget:
```
class App extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => AppState();
}

class AppState extends State<App> {

  @override
  void initState() {
    super.initState();

    Future.wait([
      _checkForPushNotificationOnLoad(),
    ]);
  }

  Future<bool> _checkForPushNotificationOnLoad() async {
    // Overwrite the handlers for when the app is in memory.
    OneSignalUtils.registerHandlers(
      communityLoader: _launchCommunityPage,
      calendarLoader: _launchCalendarPage,
    );

    // Check values from when app was first loaded.
    bool loadCommunityPage = StorageService().getLoadCommunityOnLoad();
    bool loadCalendarPage = StorageService().getLoadCalendarOnLoad();

    // Reset values.
    StorageService().setLoadCommunityOnLoad(false);
    StorageService().setLoadCalendarOnLoad(false);

    // Check to load.
    if (loadCommunityPage) {
      _launchCommunityPage();
    } else if (loadCalendarPage) {
      _launchCalendarPage();
    }

    return false;
  }

  void _launchCommunityPage() async {
    Navigator.of(context, rootNavigator: true).push(
      MaterialPageRoute(
        builder: (context) => const PageCommunity(),
      ),
    );
  }

  void _launchCalendarPage() async {
    Navigator.of(context, rootNavigator: true).push(
      MaterialPageRoute(
        builder: (context) => const PageCalendar(),
      ),
    );
  }
}
```

### OneSignalUtils:

```
import 'dart:io';
import 'package:onesignal_flutter/onesignal_flutter.dart';
import 'package:scy/services/storage.dart';

class OneSignalUtils {
  static void init() {
    if (Platform.isAndroid) {
      // // the SDK will now initialize
      // await OneSignal.shared.consentGranted(true);

      // NOTE: If this isn't working on an Android device, check that it doesn't
      // have any VPN/ad blocking software running.
    }

    OneSignal.shared.setLogLevel(OSLogLevel.verbose, OSLogLevel.none);
    OneSignal.shared.setAppId("Your app Id here");

    if (Platform.isIOS) {
      OneSignal.shared
          .promptUserForPushNotificationPermission()
          .then((accepted) {
        OneSignal.shared.disablePush(!accepted);
      });
    }

    // App is in foreground
    OneSignal.shared.setNotificationWillShowInForegroundHandler(
        (OSNotificationReceivedEvent event) {
      // Display Notification, pass null param for not displaying the notification
      event.complete(event.notification);
    });

    registerHandlers();
  }

  static void registerHandlers({
    Function? communityLoader,
    Function? calendarLoader,
  }) {
    // Notification tapped - when app hasn't loaded yet.
    OneSignal.shared
        .setNotificationOpenedHandler((OSNotificationOpenedResult result) {

      if (result.notification.additionalData != null &&
          result.notification.additionalData!.isNotEmpty) {
        result.notification.additionalData!.forEach((key, value) {
          if (key == "page" && value == "community") {
            if (communityLoader != null) {
              communityLoader();
            } else {
              StorageService().setLoadCommunityOnLoad(true);
            }
          } else if (key == "page" && value == "calendar") {
            if (calendarLoader != null) {
              calendarLoader();
            } else {
              StorageService().setLoadCalendarOnLoad(true);
            }
          }
        });
      }
    });
  }
}

```

### StorageService:

```
import 'dart:convert';
import 'package:get_storage/get_storage.dart';

class StorageService {
  static final StorageService _instance = StorageService._internal();
  final box = GetStorage();

  // using a factory is important
  // because it promises to return _an_ object of this type
  // but it doesn't promise to make a new one.
  factory StorageService() {
    return _instance;
  }

  // This named constructor is the "real" constructor
  // It'll be called exactly once, by the static property assignment above
  // it's also private, so it can only be called in this class.
  StorageService._internal() {
    // initialization logic
  }

  Future<void> clear() async {
    List<Future> list = [
      // push notifications
      box.write("loadCommunityOnLoad", null),
      box.write("loadCalendarOnLoad", null),
    ];
    await Future.wait(list);
  }


  Future<void> setLoadCommunityOnLoad(bool value) async {
    try {
      return box.write("loadCommunityOnLoad", value);
    } catch (e) {
      return Future.sync(() => null);
    }
  }

  bool getLoadCommunityOnLoad() {
    bool? data = box.read("loadCommunityOnLoad");
    if (data != null) {
      return data;
    }
    return false;
  }

  Future<void> setLoadCalendarOnLoad(bool value) async {
    try {
      return box.write("loadCalendarOnLoad", value);
    } catch (e) {
      return Future.sync(() => null);
    }
  }

  bool getLoadCalendarOnLoad() {
    bool? data = box.read("loadCalendarOnLoad");
    if (data != null) {
      return data;
    }
    return false;
  }
}
```

**Comment by daveshirman** [comment]
@alexobviously To keep this ticket cleaner, I moved my workaround over to my site, in case you still wanted it: 
[](https://www.codingandclimbing.co.uk/blog/flutter-onesignal-additionaldata-on-app-start)

**Comment by marcorubagotti** [comment]
@daveshirman i have follow your workaround as explained in your site. Like @alexobviously the handler works in all cases except when the app is closed. 
May can depend on how I have configured the routes and home settings in MaterialApp() ?

**Comment by daveshirman** [comment]
Not sure what to tell you. It works for me on both iOS and Android.

Without being able to see your code compared to mine, I couldn't possibly
help.

**Comment by alexobviously** [comment]
Thanks for your code by the way @daveshirman. I tried exactly your example with my own onesignal config and it didn't work. So that confirms that there's some very mysterious, potentially device-specific bug (although we have tested on a fairly wide range of devices and none of them worked).
The only other possibility is that we have miscommunicated at some point so I'll just clarify for anyone who comes after this (hopefully someone from onesignal):
The issue is that when apps are cold booted from a notification, i.e. they are opened from a completely closed state by tapping a onesignal notification, the handler is not called when the app opens.

**Comment by daveshirman** [comment]
So strange. I've tested this on a range of physical devices, iOS and
Android, from IPhone 6s through to 11 on the latest iOS and Android 13
even. No problems. Cold boot works as described on my post. I'm literally
using in production and all is well with several apps.

There must be something different in the setup between our code bases or
package versions maybe?

When you say you tried my example, did you literally just take my code
including the App() wrapper widget class and add some test pages instead of
PageCalendar etc?

**Comment by marcorubagotti** [comment]
@daveshirman i have reviewed your workaround and the logic work but on app closed, when i tap the notification the context or navigationKey.currentState return null when try to navigate on a simple page.

I have implemented the navigationKey and onGenerateRoute logic but not work

**Comment by daveshirman** [comment]
Just for anyone who thinks my workaround doesn't work at all, here's a video of it literally working.  This behaviour is identical on Android and this is a production app from TestFlight.

What's happening:

1. First I'm swiping up, showing you that there's no app in the recent apps list.
2. I tap the notification.
3. App loads
4. Saves a key-value store for that particular page from a payload data value in the OneSignal data
5. App checks that value and loads the appropriate page onto the stack - "Latest Content"

https://imgur.com/a/VE28Iur

**Comment by glin94** [comment]
I noticed this problem on Xiaomi Note 10s (Android 11 MIUI 12.5.9), but on another device  Xiaomi Note Pro 8 (Android 10 MIUI 12.0.5) everything works fine. I also checked on emulator (Android 11 ) and everything works

**Comment by Oliver-Knight** [comment]
@daveshirman I am also work pretty well in the same condition like you. But when I swipe and remove this notification message and didn't get this data anymore. Do u have any solution for this case.

**Comment by daveshirman** [comment]
I'm sorry, I don't understand your question.  Could you rephrase or give an
example please, so I can try to help?

**Comment by thiagotalma** [comment]
Please, someone responsible for the project, pay attention to this issue.

It is very important that it is possible to execute commands when ANY notification is received, regardless of whether the application is in foreground or background or even closed.

On Android I have this option. But I'm migrating my app to Flutter and I'm missing that option.

Flutter is becoming too relevant for the OneSignal package to be ignored.

cc:
@nan-li 
@emawby
@tanaynigam
@iAmWillShepherd
@Jeasmine
@brayden-onesignal
@gdeglin
@jkasten2
@leemunroe
@niels-s

**Comment by KKimj** [comment]
Hello!

Do I need to use another push service to route to a specific page while clicking on a push message in Terminated state?

Or has anyone else solved it?
Thank you.

**Comment by Angkhan1992** [comment]
That maybe be solution of this problem. 

```
OneSignal.Notifications.addClickListener((event) {
  logger.d('[OneSignal] addClickListener: $event');
  notification = event.notification;
  try {
    DeeplinkService.processDeeplinkUrl(
      url: event.notification.launchUrl,
    );
  } catch (e) {
    logger.e('[OneSignal] addClickListener: $e');
  }
});
```

**Comment by emawby** [comment]
Closing issue as stale. If this is still happening in 2024 feel free to tag me to reopen
